// react-prompt/src/Popup.js
import React, { Component, PropTypes } from 'react';
import wrapWithClickOut from 'react-clickout';

const propTypes = {
  closeModal: PropTypes.func,
  showModal: PropTypes.func,
  action: PropTypes.func,
  mainText: PropTypes.string,
  subText: PropTypes.string,
  actionText: PropTypes.string,
};

class Popup extends Component {
  handleClickout = () => {
    this.props.closeModal();
  }

  handleAction = () => {
    this.props.action();
  }

  handleCancel = () => {
    this.props.closeModal();
  }

  handleShow = () => {
    this.props.showModal();
  }

  render() {
    return (
      <div className="om-popup__container">
        <div className="om-popup__body">
          <div className="om-popup__header">
            {this.props.mainText}
          </div>
          <div className="om-popup__content">
            {this.props.subText}
          </div>
        </div>
        <div className="om-popup__buttons">
          <button
            className="om-btn om-btn--popup om-btn--popup-cancel"
            onClick={this.handleCancel}
          >
          Cancel
          </button>
          <button
            className="om-btn om-btn--popup om-btn--popup-delete"
            onClick={this.handleAction}
          >
          {this.props.actionText}
          </button>
        </div>
      </div>
    );
  }
}
Popup.propTypes = propTypes;

export default wrapWithClickOut(Popup);
